{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import MDAnalysis as mda\n",
    "from utils import utils\n",
    "import sbmlcore\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook generates the figures in the paper 'Predicting rifampicin resistance in M. tuberculosis using machine learning informed by protein structural and chemical features.'\n",
    "\n",
    "Authors:\n",
    "- Charlotte Lynch\n",
    "- Dylan Adlard\n",
    "- Philip Fowler"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Preperation - positions and labels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Need to create a dataset of mutations with associated phenotypes from CRypTIC tables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Data import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import data\n",
    "pheno_df = pd.read_pickle('./data/tables/cryptic/DST_MEASUREMENTS-rifamycins.pkl.gz')\n",
    "mut_df = pd.read_pickle('./data/tables/cryptic/MUTATIONS-rnap.pkl.gz')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Prepare phenotypes df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter for rifampicin phyenotypes and drop duplicates\n",
    "pheno_df = pheno_df.loc[pheno_df.DRUG=='RIF']\n",
    "pheno_df = pheno_df.drop_duplicates(subset=['UNIQUEID'], keep='first')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Prepare mutations df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter for non-synonymous solo snps in rpoB that were sequenced with FRS >= 0.9\n",
    "\n",
    "#Note, this assues minor alleles in rpoB are not clinically significant for rifampicin resistance\n",
    "mut_df = mut_df[(~mut_df['IS_NULL']) & (~mut_df['IS_SYNONYMOUS']) & (mut_df['IN_CDS'])&(mut_df['IS_FILTER_PASS'])]\n",
    "#Note, this uses the defintion of solos as variants that exist in isolation in rpoB (but can have mutations in other RNAP genes)\n",
    "mut_df = mut_df[(mut_df.IS_SNP)&(mut_df.GENE=='rpoB')]\n",
    "#Filter for solos\n",
    "mut_df.drop_duplicates(subset=['UNIQUEID'], keep=False, inplace=True)\n",
    "\n",
    "#remove phylogenetic mutations\n",
    "mut_df[~mut_df['MUTATION'].isin(['G876G', 'C66T', 'D103D', 'G86A', 'E639D'])]\n",
    "\n",
    "#insert segid column\n",
    "mut_df['segid'] = 'C'\n",
    "\n",
    "#The pdb file also attenuates the rpoB chain at position 22 and position 1147\n",
    "mut_df['resid'] = [int(i[1:-1]) for i in mut_df['MUTATION']]\n",
    "mut_df = mut_df[(mut_df['resid']>=22) & (mut_df['resid']<=1147)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Merge mutations and phenotypes dataframes\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "pheno_df.set_index('UNIQUEID', verify_integrity=True, inplace = True)\n",
    "mut_df.set_index('UNIQUEID', verify_integrity=True, inplace=True)\n",
    "merged = mut_df.join(pheno_df[['PHENOTYPE']],how='inner')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Phenotype the  mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Rather than training on all samples, we will train on unique mutations. However, many mutations appear more than once, and therefore a consensus phenotype has to be generated for these mutations to use as the label. \n",
    "\n",
    "Charlotte has decided to use an arbitrary 50% proportion of resistant samples for mutations that occur 3 or more times.\n",
    "\n",
    "Are we sure we want to do this? This is saying if a mutation is R more than 50% of the time, its classified as resistant - this doesn't make any statistical sense... And doesn't make much sense given we can generate a new, accurate catalogue for rifampicin where the variant labels are grounded in at least some logic..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>PHENOTYPE</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>segid</th>\n",
       "      <th>resid</th>\n",
       "      <th>I</th>\n",
       "      <th>R</th>\n",
       "      <th>S</th>\n",
       "      <th>U</th>\n",
       "      <th>total</th>\n",
       "      <th>prob_R</th>\n",
       "      <th>prob_S</th>\n",
       "      <th>PHENOTYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>S450L</td>\n",
       "      <td>C</td>\n",
       "      <td>450</td>\n",
       "      <td>0</td>\n",
       "      <td>5617</td>\n",
       "      <td>112</td>\n",
       "      <td>271</td>\n",
       "      <td>6000</td>\n",
       "      <td>0.936167</td>\n",
       "      <td>0.018667</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>D435V</td>\n",
       "      <td>C</td>\n",
       "      <td>435</td>\n",
       "      <td>0</td>\n",
       "      <td>726</td>\n",
       "      <td>20</td>\n",
       "      <td>29</td>\n",
       "      <td>775</td>\n",
       "      <td>0.936774</td>\n",
       "      <td>0.025806</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>H445D</td>\n",
       "      <td>C</td>\n",
       "      <td>445</td>\n",
       "      <td>0</td>\n",
       "      <td>383</td>\n",
       "      <td>8</td>\n",
       "      <td>14</td>\n",
       "      <td>405</td>\n",
       "      <td>0.945679</td>\n",
       "      <td>0.019753</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>H445Y</td>\n",
       "      <td>C</td>\n",
       "      <td>445</td>\n",
       "      <td>0</td>\n",
       "      <td>315</td>\n",
       "      <td>5</td>\n",
       "      <td>20</td>\n",
       "      <td>340</td>\n",
       "      <td>0.926471</td>\n",
       "      <td>0.014706</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>L452P</td>\n",
       "      <td>C</td>\n",
       "      <td>452</td>\n",
       "      <td>0</td>\n",
       "      <td>80</td>\n",
       "      <td>57</td>\n",
       "      <td>11</td>\n",
       "      <td>148</td>\n",
       "      <td>0.540541</td>\n",
       "      <td>0.385135</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>D997E</td>\n",
       "      <td>C</td>\n",
       "      <td>997</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>M707T</td>\n",
       "      <td>C</td>\n",
       "      <td>707</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>N135I</td>\n",
       "      <td>C</td>\n",
       "      <td>135</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>D993G</td>\n",
       "      <td>C</td>\n",
       "      <td>993</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>Y651H</td>\n",
       "      <td>C</td>\n",
       "      <td>651</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>307 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "PHENOTYPE MUTATION segid  resid  I     R    S    U  total    prob_R    prob_S  \\\n",
       "249          S450L     C    450  0  5617  112  271   6000  0.936167  0.018667   \n",
       "44           D435V     C    435  0   726   20   29    775  0.936774  0.025806   \n",
       "120          H445D     C    445  0   383    8   14    405  0.945679  0.019753   \n",
       "129          H445Y     C    445  0   315    5   20    340  0.926471  0.014706   \n",
       "166          L452P     C    452  0    80   57   11    148  0.540541  0.385135   \n",
       "..             ...   ...    ... ..   ...  ...  ...    ...       ...       ...   \n",
       "61           D997E     C    997  0     0    1    0      1  0.000000  1.000000   \n",
       "182          M707T     C    707  0     1    0    0      1  1.000000  0.000000   \n",
       "183          N135I     C    135  0     0    1    0      1  0.000000  1.000000   \n",
       "60           D993G     C    993  0     0    1    0      1  0.000000  1.000000   \n",
       "306          Y651H     C    651  0     0    1    0      1  0.000000  1.000000   \n",
       "\n",
       "PHENOTYPE PHENOTYPE  \n",
       "249               R  \n",
       "44                R  \n",
       "120               R  \n",
       "129               R  \n",
       "166               R  \n",
       "..              ...  \n",
       "61                S  \n",
       "182               R  \n",
       "183               S  \n",
       "60                S  \n",
       "306               S  \n",
       "\n",
       "[307 rows x 11 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#determine R and S occurence counts for each mutation\n",
    "ct = pd.crosstab([merged.MUTATION, merged.segid, merged.resid], merged.PHENOTYPE)\n",
    "ct[\"total\"] = ct.sum(axis=1)\n",
    "ct[\"prob_R\"] = ct[\"R\"] / ct[\"total\"]\n",
    "ct[\"prob_S\"] = ct[\"S\"] / ct[\"total\"]\n",
    "\n",
    "\n",
    "def calculate_RS(row):\n",
    "    if row[\"total\"] >= 3:\n",
    "        return \"R\" if row[\"prob_R\"] > 0.5 else \"S\" if row[\"prob_S\"] > 0.5 else \"U\"\n",
    "    return \"R\" if row[\"R\"] == row[\"total\"] else \"S\" if row[\"S\"] == row[\"total\"] else \"U\"\n",
    "\n",
    "#threshold R vs S proportion at 50% to call phenotype for mutations with n >= 3\n",
    "ct[\"PHENOTYPE\"] = ct.apply(calculate_RS, axis=1)\n",
    "ct = ct[ct[\"PHENOTYPE\"].isin([\"R\", \"S\"])]\n",
    "\n",
    "ct.reset_index(inplace=True)\n",
    "ct.sort_values(\"total\", ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Generate sample df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>segid</th>\n",
       "      <th>PHENOTYPE</th>\n",
       "      <th>resid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1002P</td>\n",
       "      <td>C</td>\n",
       "      <td>S</td>\n",
       "      <td>1002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A286V</td>\n",
       "      <td>C</td>\n",
       "      <td>R</td>\n",
       "      <td>286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A29T</td>\n",
       "      <td>C</td>\n",
       "      <td>S</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A29V</td>\n",
       "      <td>C</td>\n",
       "      <td>S</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A334D</td>\n",
       "      <td>C</td>\n",
       "      <td>S</td>\n",
       "      <td>334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302</th>\n",
       "      <td>V895F</td>\n",
       "      <td>C</td>\n",
       "      <td>S</td>\n",
       "      <td>895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>V946G</td>\n",
       "      <td>C</td>\n",
       "      <td>S</td>\n",
       "      <td>946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>Y308C</td>\n",
       "      <td>C</td>\n",
       "      <td>R</td>\n",
       "      <td>308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>Y572C</td>\n",
       "      <td>C</td>\n",
       "      <td>S</td>\n",
       "      <td>572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>Y651H</td>\n",
       "      <td>C</td>\n",
       "      <td>S</td>\n",
       "      <td>651</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>307 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    mutation segid PHENOTYPE  resid\n",
       "0     A1002P     C         S   1002\n",
       "1      A286V     C         R    286\n",
       "2       A29T     C         S     29\n",
       "3       A29V     C         S     29\n",
       "4      A334D     C         S    334\n",
       "..       ...   ...       ...    ...\n",
       "302    V895F     C         S    895\n",
       "303    V946G     C         S    946\n",
       "304    Y308C     C         R    308\n",
       "305    Y572C     C         S    572\n",
       "306    Y651H     C         S    651\n",
       "\n",
       "[307 rows x 4 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create final DataFrame from ct\n",
    "df = pd.DataFrame({\n",
    "    'mutation': ct.MUTATION,\n",
    "    'segid': ct['segid'],\n",
    "    'PHENOTYPE': ct['PHENOTYPE'],\n",
    "    'resid': ct['resid']\n",
    "})\n",
    "\n",
    "# Optionally reset index if needed (though it may not be necessary)\n",
    "df.reset_index(drop=True, inplace=True)\n",
    "df.to_csv('./data/tables/generated/rpoB_solos.csv')\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Proportion of Resisant mutations: PHENOTYPE\n",
      "S    259\n",
      "R     48\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# R vs S weighting:\n",
    "print ('Proportion of Resisant mutations:', df.PHENOTYPE.value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This dataframe basically acts as a catalogue, with unique mutations and their associated labels."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutation mapping onto RNAP and rpoB structures"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Where on the structure of rpoB do these mutation exist?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/dylanadlard/miniforge3/envs/rif_manuscript/lib/python3.12/site-packages/MDAnalysis/coordinates/PDB.py:1153: UserWarning: Found no information for attr: 'formalcharges' Using default value of '0'\n",
      "  warnings.warn(\"Found no information for attr: '{}'\"\n",
      "/Users/dylanadlard/miniforge3/envs/rif_manuscript/lib/python3.12/site-packages/MDAnalysis/coordinates/PDB.py:1153: UserWarning: Found no information for attr: 'formalcharges' Using default value of '0'\n",
      "  warnings.warn(\"Found no information for attr: '{}'\"\n"
     ]
    }
   ],
   "source": [
    "utils.map_mut2pdb(df, './data/pdb/5uh6.pdb', './data/pdb/rnap_mutations.pdb')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TO DO: generate figure 1 and display the png here."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data preperation - Feature engineering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have a table of mutations and their labels. Next step is to generate a global feature set for each mutation using sbmlcore."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "features = sbmlcore.FeatureDataset(df, protein=\"RNAP\", species=\"M. tuberculosis\", gene=\"rpoB\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Generate amino acid features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "#change in amino acid volume on mutation\n",
    "volume = sbmlcore.AminoAcidVolumeChange()\n",
    "#change in amino acid hydropathy on mutation\n",
    "hydropathy = sbmlcore.AminoAcidHydropathyChangeKyteDoolittle()\n",
    "#change in amino acid molecular weight on mutation\n",
    "mw = sbmlcore.AminoAcidMWChange()\n",
    "#change in amino acid isoelectric point on mutation\n",
    "pi = sbmlcore.AminoAcidPiChange()\n",
    "#change in residue-environment similarity\n",
    "rogov = sbmlcore.AminoAcidRogovChange()\n",
    "\n",
    "features.add_feature([volume, hydropathy, mw, pi, rogov])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Generate distance-based features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "#distance from each residue to rifampicin\n",
    "rif_distance = sbmlcore.StructuralDistances('./data/pdb/5uh6.pdb','resname RFP', 'Rif_distance', infer_masses=True, offsets = {\"C\": -6})\n",
    "#distance from each residue to the magnesium ion\n",
    "mg_distance = sbmlcore.StructuralDistances('./data/pdb/5uh6.pdb','resname MG', 'Mg_distance', infer_masses=True, offsets = {\"C\": -6})\n",
    "#distance from each residue to the first zinc ion\n",
    "zn1_distance = sbmlcore.StructuralDistances('./data/pdb/5uh6.pdb','index 26082 and resname ZN', 'Zn1_distance', infer_masses=True, offsets = {\"C\": -6})\n",
    "#distance from each residue to the second zinc ion\n",
    "zn2_distance = sbmlcore.StructuralDistances('./data/pdb/5uh6.pdb','index 26083 and resname ZN', 'Zn2_distance', infer_masses=True, offsets = {\"C\": -6})\n",
    "#distance from each residue to the antisense DNA strand\n",
    "antisense_p_distance = sbmlcore.StructuralDistances('./data/pdb/5uh6.pdb','segid G and name P', 'antisense_P_distance', infer_masses=True, offsets = {\"C\": -6})\n",
    "#distance from each residue to the coding DNA strand\n",
    "sense_p_distance = sbmlcore.StructuralDistances('./data/pdb/5uh6.pdb','segid H and name P', 'sense_P_distance', infer_masses=True, offsets = {\"C\": -6})\n",
    "#distance from each residue to the crystallised mRNA molecule\n",
    "rna_distance = sbmlcore.StructuralDistances('./data/pdb/5uh6.pdb','segid I', 'RNA_distance', infer_masses=True, offsets = {\"C\": -6})\n",
    "\n",
    "features.add_feature([rif_distance, mg_distance, zn1_distance, zn2_distance, antisense_p_distance, sense_p_distance, rna_distance])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####  Generate structure-related features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/dylanadlard/miniforge3/envs/rif_manuscript/lib/python3.12/site-packages/sbmlcore/ExternalCode.py:313: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.\n",
      "The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.\n",
      "\n",
      "For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.\n",
      "\n",
      "\n",
      "  snap2_df['Expected Accuracy'].replace(to_replace='%', value='', regex=True, inplace=True)\n"
     ]
    }
   ],
   "source": [
    "# use stride to assign secondary structure to each residue\n",
    "stride = sbmlcore.Stride(\"./data/pdb/5uh6-peptide-only.pdb\", offsets={'A': 0, 'B': 0, 'C':-6, 'D':0, 'E':0, 'F':0})\n",
    "# use freeSASA to calculate solvent accessible surface area of each residue\n",
    "freesasa = sbmlcore.FreeSASA(\"./data/pdb/5uh6.pdb\", offsets = {'C':-6})\n",
    "# use SNAP2 to predict functional perturbation for each mutation (in reality just adding pre-generated feature to df)\n",
    "snap2 = sbmlcore.SNAP2(\"./data/stride/5uh6-complete.csv\", offsets={'A': 0, 'B': 0, 'C':-6, 'D':0, 'E':0, 'F':0})\n",
    "# use deepddg to predict change in stability on mutation\n",
    "deepddg = sbmlcore.DeepDDG(\"./data/ddg/5uh6.ddg\", offsets={'A': 0, 'B': 0, 'C':-6, 'D':0, 'E':0, 'F':0})\n",
    "# use rasp to predict change in stability on mutation\n",
    "rasp = sbmlcore.RaSP(\"./data/rasp/cavity_pred_5uh6_C.csv\", offsets = {'C':-6})\n",
    "# extract temperature factors from pdb file and add to df\n",
    "temp = sbmlcore.TempFactors(\"./data/pdb/5uh6.pdb\", offsets={'A': 0, 'B': 0, 'C':-6, 'D':0, 'E':0, 'F':0})\n",
    "\n",
    "features.add_feature([stride, freesasa, snap2, deepddg, rasp, temp])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Clean up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>segid</th>\n",
       "      <th>mutation</th>\n",
       "      <th>PHENOTYPE</th>\n",
       "      <th>d_volume</th>\n",
       "      <th>d_hydropathy_KD</th>\n",
       "      <th>d_MW</th>\n",
       "      <th>d_pi</th>\n",
       "      <th>d_rogov</th>\n",
       "      <th>Rif_distance</th>\n",
       "      <th>Mg_distance</th>\n",
       "      <th>...</th>\n",
       "      <th>phi</th>\n",
       "      <th>psi</th>\n",
       "      <th>n_hbond_acceptors</th>\n",
       "      <th>n_hbond_donors</th>\n",
       "      <th>SASA</th>\n",
       "      <th>snap2_score</th>\n",
       "      <th>deep_ddG</th>\n",
       "      <th>rasp_score</th>\n",
       "      <th>temp_factor</th>\n",
       "      <th>secondary_structure</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>A1002P</td>\n",
       "      <td>S</td>\n",
       "      <td>24.1</td>\n",
       "      <td>-3.4</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.020</td>\n",
       "      <td>45.415055</td>\n",
       "      <td>31.783345</td>\n",
       "      <td>...</td>\n",
       "      <td>-155.77</td>\n",
       "      <td>164.84</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.268116</td>\n",
       "      <td>36</td>\n",
       "      <td>-2.927</td>\n",
       "      <td>0.800033</td>\n",
       "      <td>36.009998</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>A286V</td>\n",
       "      <td>R</td>\n",
       "      <td>51.4</td>\n",
       "      <td>2.4</td>\n",
       "      <td>28.0</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>0.232</td>\n",
       "      <td>47.831892</td>\n",
       "      <td>64.028439</td>\n",
       "      <td>...</td>\n",
       "      <td>-98.83</td>\n",
       "      <td>-17.70</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-31</td>\n",
       "      <td>-0.523</td>\n",
       "      <td>0.222401</td>\n",
       "      <td>121.500000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>A29T</td>\n",
       "      <td>S</td>\n",
       "      <td>27.5</td>\n",
       "      <td>-2.5</td>\n",
       "      <td>30.0</td>\n",
       "      <td>-0.40</td>\n",
       "      <td>0.248</td>\n",
       "      <td>34.306893</td>\n",
       "      <td>40.030977</td>\n",
       "      <td>...</td>\n",
       "      <td>-91.15</td>\n",
       "      <td>161.38</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>34.753706</td>\n",
       "      <td>-81</td>\n",
       "      <td>-1.142</td>\n",
       "      <td>0.328659</td>\n",
       "      <td>32.820000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>A29V</td>\n",
       "      <td>S</td>\n",
       "      <td>51.4</td>\n",
       "      <td>2.4</td>\n",
       "      <td>28.0</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>0.232</td>\n",
       "      <td>34.306893</td>\n",
       "      <td>40.030977</td>\n",
       "      <td>...</td>\n",
       "      <td>-91.15</td>\n",
       "      <td>161.38</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>34.753706</td>\n",
       "      <td>-76</td>\n",
       "      <td>-1.056</td>\n",
       "      <td>0.284229</td>\n",
       "      <td>32.820000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C</td>\n",
       "      <td>A334D</td>\n",
       "      <td>S</td>\n",
       "      <td>22.5</td>\n",
       "      <td>-5.3</td>\n",
       "      <td>44.0</td>\n",
       "      <td>-3.23</td>\n",
       "      <td>0.029</td>\n",
       "      <td>43.228183</td>\n",
       "      <td>56.772771</td>\n",
       "      <td>...</td>\n",
       "      <td>-67.17</td>\n",
       "      <td>-16.26</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.786252</td>\n",
       "      <td>-54</td>\n",
       "      <td>-2.734</td>\n",
       "      <td>0.369132</td>\n",
       "      <td>102.089996</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302</th>\n",
       "      <td>C</td>\n",
       "      <td>V895F</td>\n",
       "      <td>S</td>\n",
       "      <td>49.9</td>\n",
       "      <td>-1.4</td>\n",
       "      <td>48.1</td>\n",
       "      <td>-0.48</td>\n",
       "      <td>0.345</td>\n",
       "      <td>41.296670</td>\n",
       "      <td>28.986757</td>\n",
       "      <td>...</td>\n",
       "      <td>-59.98</td>\n",
       "      <td>-22.89</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>35.841417</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1.014</td>\n",
       "      <td>0.287163</td>\n",
       "      <td>37.750000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>C</td>\n",
       "      <td>V946G</td>\n",
       "      <td>S</td>\n",
       "      <td>-79.9</td>\n",
       "      <td>-4.6</td>\n",
       "      <td>-42.0</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-0.080</td>\n",
       "      <td>53.471656</td>\n",
       "      <td>48.889279</td>\n",
       "      <td>...</td>\n",
       "      <td>-77.65</td>\n",
       "      <td>122.58</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86.193134</td>\n",
       "      <td>-81</td>\n",
       "      <td>-0.444</td>\n",
       "      <td>0.424508</td>\n",
       "      <td>68.440002</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>C</td>\n",
       "      <td>Y308C</td>\n",
       "      <td>R</td>\n",
       "      <td>-85.1</td>\n",
       "      <td>3.8</td>\n",
       "      <td>-60.0</td>\n",
       "      <td>-0.59</td>\n",
       "      <td>-0.199</td>\n",
       "      <td>33.128330</td>\n",
       "      <td>49.415429</td>\n",
       "      <td>...</td>\n",
       "      <td>-65.00</td>\n",
       "      <td>-35.81</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>70.874065</td>\n",
       "      <td>68</td>\n",
       "      <td>-1.361</td>\n",
       "      <td>0.467171</td>\n",
       "      <td>76.190002</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>C</td>\n",
       "      <td>Y572C</td>\n",
       "      <td>S</td>\n",
       "      <td>-85.1</td>\n",
       "      <td>3.8</td>\n",
       "      <td>-60.0</td>\n",
       "      <td>-0.59</td>\n",
       "      <td>-0.199</td>\n",
       "      <td>37.911061</td>\n",
       "      <td>45.458752</td>\n",
       "      <td>...</td>\n",
       "      <td>-129.88</td>\n",
       "      <td>172.92</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>27.871431</td>\n",
       "      <td>35</td>\n",
       "      <td>-1.543</td>\n",
       "      <td>0.563006</td>\n",
       "      <td>61.090000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>C</td>\n",
       "      <td>Y651H</td>\n",
       "      <td>S</td>\n",
       "      <td>-40.4</td>\n",
       "      <td>-1.9</td>\n",
       "      <td>-26.0</td>\n",
       "      <td>1.93</td>\n",
       "      <td>0.073</td>\n",
       "      <td>35.306373</td>\n",
       "      <td>45.340051</td>\n",
       "      <td>...</td>\n",
       "      <td>-152.68</td>\n",
       "      <td>170.72</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>76.740001</td>\n",
       "      <td>-41</td>\n",
       "      <td>-0.727</td>\n",
       "      <td>0.318788</td>\n",
       "      <td>29.900000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>307 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    segid mutation PHENOTYPE  d_volume  d_hydropathy_KD  d_MW  d_pi  d_rogov  \\\n",
       "0       C   A1002P         S      24.1             -3.4  26.0  0.30    0.020   \n",
       "1       C    A286V         R      51.4              2.4  28.0 -0.04    0.232   \n",
       "2       C     A29T         S      27.5             -2.5  30.0 -0.40    0.248   \n",
       "3       C     A29V         S      51.4              2.4  28.0 -0.04    0.232   \n",
       "4       C    A334D         S      22.5             -5.3  44.0 -3.23    0.029   \n",
       "..    ...      ...       ...       ...              ...   ...   ...      ...   \n",
       "302     C    V895F         S      49.9             -1.4  48.1 -0.48    0.345   \n",
       "303     C    V946G         S     -79.9             -4.6 -42.0  0.01   -0.080   \n",
       "304     C    Y308C         R     -85.1              3.8 -60.0 -0.59   -0.199   \n",
       "305     C    Y572C         S     -85.1              3.8 -60.0 -0.59   -0.199   \n",
       "306     C    Y651H         S     -40.4             -1.9 -26.0  1.93    0.073   \n",
       "\n",
       "     Rif_distance  Mg_distance  ...     phi     psi  n_hbond_acceptors  \\\n",
       "0       45.415055    31.783345  ... -155.77  164.84                1.0   \n",
       "1       47.831892    64.028439  ...  -98.83  -17.70                0.0   \n",
       "2       34.306893    40.030977  ...  -91.15  161.38                0.0   \n",
       "3       34.306893    40.030977  ...  -91.15  161.38                0.0   \n",
       "4       43.228183    56.772771  ...  -67.17  -16.26                1.0   \n",
       "..            ...          ...  ...     ...     ...                ...   \n",
       "302     41.296670    28.986757  ...  -59.98  -22.89                0.0   \n",
       "303     53.471656    48.889279  ...  -77.65  122.58                0.0   \n",
       "304     33.128330    49.415429  ...  -65.00  -35.81                1.0   \n",
       "305     37.911061    45.458752  ... -129.88  172.92                1.0   \n",
       "306     35.306373    45.340051  ... -152.68  170.72                1.0   \n",
       "\n",
       "     n_hbond_donors       SASA  snap2_score  deep_ddG  rasp_score  \\\n",
       "0               1.0   7.268116           36    -2.927    0.800033   \n",
       "1               1.0   0.000000          -31    -0.523    0.222401   \n",
       "2               0.0  34.753706          -81    -1.142    0.328659   \n",
       "3               0.0  34.753706          -76    -1.056    0.284229   \n",
       "4               1.0   0.786252          -54    -2.734    0.369132   \n",
       "..              ...        ...          ...       ...         ...   \n",
       "302             0.0  35.841417           -1    -1.014    0.287163   \n",
       "303             0.0  86.193134          -81    -0.444    0.424508   \n",
       "304             1.0  70.874065           68    -1.361    0.467171   \n",
       "305             1.0  27.871431           35    -1.543    0.563006   \n",
       "306             0.0  76.740001          -41    -0.727    0.318788   \n",
       "\n",
       "     temp_factor  secondary_structure  \n",
       "0      36.009998                    0  \n",
       "1     121.500000                    1  \n",
       "2      32.820000                    2  \n",
       "3      32.820000                    2  \n",
       "4     102.089996                    1  \n",
       "..           ...                  ...  \n",
       "302    37.750000                    3  \n",
       "303    68.440002                    2  \n",
       "304    76.190002                    1  \n",
       "305    61.090000                    0  \n",
       "306    29.900000                    0  \n",
       "\n",
       "[307 rows x 25 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#annotate secondary structures with integer codes\n",
    "features.df['secondary_structure_codes'] = pd.Categorical(features.df.secondary_structure, categories=features.df.secondary_structure.unique()).codes\n",
    "features.df = features.df.drop(columns=[\"secondary_structure\",\"secondary_structure_long\",\"B\",\"C\",\"E\",\"G\",\"H\",\"T\"])\n",
    "#drop superfluous columns\n",
    "features.df = features.df.drop(columns=[\"residue_sasa\",\"snap2_accuracy\",\"rasp_wt_nlf\",\"rasp_mt_nlf\",\"rasp_score_ml\"])\n",
    "#rename columns\n",
    "features.df = features.df.rename(columns={\"d_Pi\":\"d_pi\", \"rasp_score_ml_fermi\":\"rasp_score\", \"secondary_structure_codes\":\"secondary_structure\"})\n",
    "\n",
    "features.df.to_csv('./data/tables/generated/features_dataset.csv')\n",
    "df = features.df\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "BDQ_analysis",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
